core/templates/dev/head/filters.js

```diff
@@ -240,6 +240,15 @@ oppia.filter('parameterizeRuleDescription', [
           replacementText = inputs[varName];
         }
 
+        if (typeof (replacementText) === 'string') {
```

Comment:
This seems quite hacky. Why don't you do it based on vartype?

Reply:
will try that out

Reply:
Actually this issue could occur for any rule which is passed through the filter. That's why I feel a more general check to make sure $ signs are properly handled (and not caught errenously by the matching in the string replace call that follows this block). So I think the best way out is to apply this on every string.. Any better suggestions?

Reply:
Oh, I see -- this acts on the derived string. In that case, my question changes: why is replacementText not always a string?

Reply:
It is not a string in the case of NumericInput. In all the separately handled cases it is explicitly set to a string except in the final else statement https://github.com/oppia/oppia/blob/6e07cd76df57e3448494696d8bbf6244a0e6c0b7/core/templates/dev/head/filters.js#L239-L241. Should I force cast replacementText to a string instead even in the else block? As finalDescription (the value thats returned) is always a string?

Reply:
For NumericInput I'd convert it to a string. For the else case could you audit all the interactions and change the logic to something like "else if (varType !== 'string') throw Error"? That way we can be sure that every case is covered (and continues to be covered).

Then you don't need the type check for further modifications of replacementText.

Reply:
I removed this check. I manually checked all the interaction rules using all_interactions exploration. only set input and numeric input had to be managed separately. This should guarantee that replacement text is always a string